pylds/base.py

```diff
@@ -6,32 +6,122 @@
-    Y, X = np.meshgrid(points_y, points_x)  # Grid in phase-space
-    mesh = np.transpose([X.flatten(), Y.flatten(), np.zeros(Nx*Ny)]) # 2D grid + a zero column for LDs
-    return mesh.flatten()
+    if type(grid_parameters) == dict:
+        # Unpack extra grid parameters
+        slice_parameters = grid_parameters['slice_parameters']
+        dof_slice = grid_parameters['dof_slice']
+        dof_fixed, dof_fixed_values = grid_parameters['dof_fixed']
+        momentum_sign = grid_parameters['momentum_sign']
+        potential_energy = grid_parameters['potential_energy']
+        H0 = grid_parameters['energy_level']
+
+        N_dim = len(dof_slice)  # Total number of DoF
+        # Check N DoF is even.
+        if N_dim % 2 != 0:
+            error_mssg = ("ERROR: Number of DoF not even. ",
+                          "Check your extra grid parameters")
+            print(error_mssg)
+            sys.exit()
+        # Determine number of DoF for Energy conservation.
+        # There must be only one.
+        dof_remaining = 1 - (np.array(dof_fixed) + np.array(dof_slice))
+        if list(dof_remaining).count(1) > 1:
+            error_mssg = ("ERROR: More than one remaing DoF. ",
+                          "Cannot use Energy conservation to define high-dim grid.")
+            print(error_mssg)
+            sys.exit()
+
+        # Axes of visualisation slice
+        def pass_axis_parameters(parameters): return np.linspace(*parameters)
+        points_slice_axes = list(map(pass_axis_parameters, slice_parameters))
+        slice_mesh = np.meshgrid(*points_slice_axes)
+        dof_slice_axes = [axis.flatten() for axis in slice_mesh]
+
+        # Axes of fixed DoF
+        dof_fixed_axes = dof_fixed_values
+
+        N_points_slice = np.prod(list(map(itemgetter(-1), slice_parameters)))
+        phase_space_axes = {}
+
+        # Define and sort axes in phase space
+        k = 0
+        for i in range(N_dim):
+            if dof_slice[i] == 1:
+                phase_space_axes[i] = dof_slice_axes[k]
+                k += 1
+        k = 0
+        for i in range(N_dim):
+            if dof_fixed[i] == 1:
+                phase_space_axes[i] = dof_fixed_axes[k] * \
+                    np.zeros(N_points_slice)
+                k += 1
+
+        # Set axis to be determined by energy conservation
+        idx_dof_H0 = list(set(range(N_dim))-set(phase_space_axes.keys()))[0]
+        phase_space_axes[idx_dof_H0] = np.zeros(N_points_slice)
+
+        # List of all phase space axes
+        phase_space_axes = [phase_space_axes[i] for i in range(int(N_dim))]
+
+        # Determine undefined axis via energy conservation
+        phase_space_axes[idx_dof_H0] = energy_conservation_condition(
+            phase_space_axes, H0, potential_energy, momentum_sign)
+        
+        mask = np.isnan(phase_space_axes[idx_dof_H0]) # Mask grid points
+        phase_space_axes[idx_dof_H0] = np.nan_to_num(phase_space_axes[idx_dof_H0])
```

Comment:
I'd think it's safer to leave keep the nans rather than having artificial zeros that do not satisy the energy condition

Reply:
Yes, that was a change that I had already implemented. I don't know why you're seeing them.
Did you pull my changes from GitHub before you started modifying the code?

Reply:
Hi @vkrajnak 

I see. 

Yes, I tried to run the code leaving `phase_space_axes[idx_dof_H0]` with `Nan` values. However, once I executed `compute_lagrangian_descriptor` although it didn't throw any immediate error, the run was just taking too long - I tried this a couple of times. And once, I changed values to zero using `np.nan_to_num`, it ran fine. Then, I just used the `mask` values to get the desired output. 

Maybe there's an explanation to this that I couldn't see. Try it yourself if you feel like to convince yourself.

Reply:
`solve_ivp` is written in such a way, that it will run with all nans in an infinite loop. Applying the mask to initial conditions, as you did below, sounds a lot safer